File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map
  def count(sentence) do
    sentence
    |> String.downcase
    |> String.split(~r{[\s,_:!&@$%^]+}, trim: true)
    |> Enum.reduce(%{}, fn(word, dic) -> update_dictionary(dic, word) end)
  end

  defp update_dictionary(dictionary, word) do
    count = Map.get(dictionary, word) || 0
    Map.put(dictionary, word, count + 1)
  end
end

File: clojure/word-count/src/word_count.clj
(ns word-count
  (:require [clojure.string :refer [split lower-case]]))

(defn word-count [sentence]
  (reduce
    #(assoc %1 %2 (inc (%1 %2 0)))
    {}
    (split (lower-case sentence) #"\W+"))
)
